File: reference-implementations/typescript/BlueprintCompiler.ts
/**
 * BlueprintCompiler - NORP-005 Reference Implementation
 *
 * Implements:
 * - NORP-005: Deterministic Topological Ordering (Kahn's algorithm)
 * - NORP-004: Cycle detection via topological sort
 *
 * @license MIT
 * @copyright 2026 NeuraScope CONVERWAY
 */

import { ExecutionPlanImpl } from './ExecutionPlan';
import type { ExecutionPlan, Node, ParallelGroup } from './types';

export class BlueprintCompiler {
  /**
   * Compile workflow into deterministic execution plan
   */
  compile(workflow: { nodes: Node[] }): ExecutionPlan {
    const nodes = workflow.nodes || [];

    if (nodes.length === 0) {
      throw new Error('No nodes to compile');
    }

    // 1. Build dependency graph
    const graph = this.buildDependencyGraph(nodes);

    // 2. Topological sort (NORP-005)
    const executionOrder = this.topologicalSort(nodes, graph);

    // 3. Detect parallelizable groups
    const parallelGroups = this.detectParallelGroups(graph);

    // 4. Estimate duration
    const estimatedDuration = this.estimateDuration(nodes);

    return new ExecutionPlanImpl(
      nodes,
      executionOrder,
      parallelGroups,
      estimatedDuration
    );
  }

  /**
   * Topological sort using Kahn's algorithm (NORP-005)
   * Complexity: O(V + E)
   */
  private topologicalSort(
    nodes: Node[],
    graph: Record<string, string[]>
  ): string[] {
    // 1. Calculate in-degree for each node
    const inDegree: Record<string, number> = {};

    for (const node of nodes) {
      inDegree[node.id] = 0;
    }

    for (const node of nodes) {
      const dependencies = node.depends_on || [];

      for (const depId of dependencies) {
        if (depId in inDegree) {
          inDegree[node.id]++;
        }
      }
    }

    // 2. Queue with zero in-degree nodes
    // NORP-005: Deterministic tie-breaking (sorted)
    let queue = Object.keys(inDegree)
      .filter(nodeId => inDegree[nodeId] === 0)
      .sort();

    // 3. BFS processing
    const result: string[] = [];

    while (queue.length > 0) {
      const current = queue.shift()!;
      result.push(current);

      const newlyEligible: string[] = [];

      // Find nodes that depend on current
      for (const node of nodes) {
        if (node.depends_on?.includes(current)) {
          inDegree[node.id]--;

          if (inDegree[node.id] === 0) {
            newlyEligible.push(node.id);
          }
        }
      }

      // NORP-005: Deterministic reinsertion (sorted)
      if (newlyEligible.length > 0) {
        queue = [...queue, ...newlyEligible].sort();
      }
    }

    // 4. Verify all nodes sorted (else cycle)
    if (result.length !== nodes.length) {
      throw new Error(
        `Compilation failed: Cycle detected in graph. ` +
        `Nodes sorted: ${result.length}/${nodes.length}`
      );
    }

    return result;
  }

  /**
   * Build dependency graph
   */
  private buildDependencyGraph(nodes: Node[]): Record<string, string[]> {
    const graph: Record<string, string[]> = {};

    for (const node of nodes) {
      graph[node.id] = node.depends_on || [];
    }

    return graph;
  }

  /**
   * Detect parallelizable groups (NORP-005)
   */
  private detectParallelGroups(graph: Record<string, string[]>): ParallelGroup[] {
    const levels = this.computeLevels(graph);
    const groups: ParallelGroup[] = [];

    for (const [level, nodeIds] of Object.entries(levels)) {
      groups.push({
        level: parseInt(level),
        nodes: Object.freeze(nodeIds),
        parallel: nodeIds.length > 1
      });
    }

    return groups;
  }

  /**
   * Compute dependency levels (BFS)
   */
  private computeLevels(graph: Record<string, string[]>): Record<number, string[]> {
    const levels: Record<number, string[]> = {};
    const nodeLevel: Record<string, number> = {};
    const queue: [string, number][] = [];

    // Init: Nodes without dependencies = level 0
    for (const [nodeId, dependencies] of Object.entries(graph)) {
      if (dependencies.length === 0) {
        queue.push([nodeId, 0]);
        nodeLevel[nodeId] = 0;
      }
    }

    // BFS
    while (queue.length > 0) {
      const [current, level] = queue.shift()!;

      if (!levels[level]) {
        levels[level] = [];
      }
      levels[level].push(current);

      // Find nodes that depend on current
      for (const [nodeId, dependencies] of Object.entries(graph)) {
        if (dependencies.includes(current) && !(nodeId in nodeLevel)) {
          // Calculate level = max(dependencies levels) + 1
          const maxDepLevel = Math.max(
            ...dependencies.map(depId => nodeLevel[depId] || 0)
          );

          nodeLevel[nodeId] = maxDepLevel + 1;
          queue.push([nodeId, maxDepLevel + 1]);
        }
      }
    }

    return levels;
  }

  /**
   * Estimate total execution duration
   */
  private estimateDuration(nodes: Node[]): number {
    const durationByType: Record<string, number> = {
      'datasource': 200,
      'llm_call': 2000,
      'custom_code': 100,
      'conditional': 5,
      'loop': 500,
      'output': 50,
    };

    return nodes.reduce((total, node) => {
      return total + (durationByType[node.type] || 100);
    }, 0);
  }
}

File: reference-implementations/python/blueprint_compiler.py
"""
BlueprintCompiler - NORP-005 Reference Implementation

Implements:
- NORP-005: Deterministic Topological Ordering (Kahn's algorithm)
- NORP-004: Cycle detection via topological sort

License: MIT
Copyright: 2026 NeuraScope CONVERWAY
"""

from typing import List, Dict, Any
from execution_plan import ExecutionPlan


class BlueprintCompiler:
    """
    NORP-compliant workflow compiler
    """

    def compile(self, workflow: dict) -> ExecutionPlan:
        """
        Compile workflow into deterministic execution plan

        Args:
            workflow: Workflow with 'nodes' array

        Returns:
            ExecutionPlan (immutable DTO)

        Raises:
            Exception: If cycle detected
        """
        nodes = workflow.get('nodes', [])

        if not nodes:
            raise Exception('No nodes to compile')

        # 1. Build dependency graph
        graph = self._build_dependency_graph(nodes)

        # 2. Topological sort (NORP-005)
        execution_order = self._topological_sort(nodes, graph)

        # 3. Detect parallelizable groups
        parallel_groups = self._detect_parallel_groups(graph)

        # 4. Estimate duration
        estimated_duration = self._estimate_duration(nodes)

        return ExecutionPlan(
            nodes=nodes,
            execution_order=execution_order,
            parallel_groups=parallel_groups,
            estimated_duration_ms=estimated_duration
        )

    def _topological_sort(self, nodes: List[dict], graph: Dict[str, List[str]]) -> List[str]:
        """
        Topological sort using Kahn's algorithm (NORP-005)

        Complexity: O(V + E)

        Args:
            nodes: List of nodes
            graph: Dependency graph

        Returns:
            Deterministic execution order (node IDs)

        Raises:
            Exception: If cycle detected
        """
        # 1. Calculate in-degree for each node
        in_degree = {}

        for node in nodes:
            node_id = node['id']
            in_degree[node_id] = 0

        for node in nodes:
            node_id = node['id']
            dependencies = node.get('depends_on', [])

            for dep_id in dependencies:
                if dep_id in in_degree:
                    in_degree[node_id] += 1

        # 2. Queue with zero in-degree nodes
        # NORP-005: Deterministic tie-breaking (sorted)
        queue = sorted([
            node_id
            for node_id, degree in in_degree.items()
            if degree == 0
        ])

        # 3. BFS processing
        result = []

        while queue:
            current = queue.pop(0)
            result.append(current)

            newly_eligible = []

            # Find nodes that depend on current
            for node in nodes:
                node_id = node['id']

                if current in node.get('depends_on', []):
                    in_degree[node_id] -= 1

                    if in_degree[node_id] == 0:
                        newly_eligible.append(node_id)

            # NORP-005: Deterministic reinsertion (sorted)
            if newly_eligible:
                queue.extend(newly_eligible)
                queue.sort()

        # 4. Verify all nodes sorted (else cycle)
        if len(result) != len(nodes):
            raise Exception(
                f'Compilation failed: Cycle detected in graph. '
                f'Nodes sorted: {len(result)}/{len(nodes)}'
            )

        return result

    def _build_dependency_graph(self, nodes: List[dict]) -> Dict[str, List[str]]:
        """Build dependency graph"""
        graph = {}

        for node in nodes:
            node_id = node['id']
            graph[node_id] = node.get('depends_on', [])

        return graph

    def _detect_parallel_groups(self, graph: Dict[str, List[str]]) -> List[dict]:
        """
        Detect parallelizable groups (NORP-005)

        Args:
            graph: Dependency graph

        Returns:
            List of parallel groups with levels
        """
        levels = self._compute_levels(graph)
        groups = []

        for level, node_ids in levels.items():
            groups.append({
                'level': level,
                'nodes': node_ids,
                'parallel': len(node_ids) > 1
            })

        return groups

    def _compute_levels(self, graph: Dict[str, List[str]]) -> Dict[int, List[str]]:
        """
        Compute dependency levels (BFS)

        Args:
            graph: Dependency graph

        Returns:
            {'level': ['node1', 'node2']}
        """
        levels = {}
        node_level = {}
        queue = []

        # Init: Nodes without dependencies = level 0
        for node_id, dependencies in graph.items():
            if not dependencies:
                queue.append((node_id, 0))
                node_level[node_id] = 0

        # BFS
        while queue:
            current, level = queue.pop(0)

            if level not in levels:
                levels[level] = []
            levels[level].append(current)

            # Find nodes that depend on current
            for node_id, dependencies in graph.items():
                if current in dependencies and node_id not in node_level:
                    # Calculate level = max(dependencies levels) + 1
                    max_dep_level = max(
                        (node_level.get(dep_id, 0) for dep_id in dependencies),
                        default=0
                    )

                    node_level[node_id] = max_dep_level + 1
                    queue.append((node_id, max_dep_level + 1))

        return levels

    def _estimate_duration(self, nodes: List[dict]) -> int:
        """
        Estimate total execution duration

        Args:
            nodes: List of nodes

        Returns:
            Duration in milliseconds
        """
        duration_by_type = {
            'datasource': 200,
            'llm_call': 2000,
            'custom_code': 100,
            'conditional': 5,
            'loop': 500,
            'output': 50,
        }

        total_duration = sum(
            duration_by_type.get(node.get('type', 'unknown'), 100)
            for node in nodes
        )

        return total_duration
